In [24]:
from langchain.output_parsers.json import SimpleJsonOutputParser
from langchain.prompts import PromptTemplate
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.messages import HumanMessage, SystemMessage 
from dotenv import load_dotenv
from datetime import datetime
from os.path import exists
import tiktoken
import re
import os
import instructor
from openai import OpenAI
from pydantic import BaseModel, Field
from typing import List, Optional, Literal
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_openai import ChatOpenAI
from langchain.output_parsers.json import SimpleJsonOutputParser
from langchain_core.runnables.history import RunnableWithMessageHistory
from langchain_community.chat_message_histories import ChatMessageHistory
from langchain.schema import StrOutputParser
import json
from tqdm import tqdm


In [13]:
def read_jsonl_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        return [json.loads(line) for line in file if line.strip()]
    

def write_jsonl_file(file_path, data):
    # Open the file in write mode ('w')
    with open(file_path, 'w') as file:
        for item in data:
            # Serialize the dictionary to a JSON formatted string
            json_str = json.dumps(item)
            # Write the JSON string to the file followed by a newline
            file.write(json_str + '\n')


In [73]:
data = read_jsonl_file('final_target_scores.jsonl')

In [59]:
jewelry = [product for product in data if product['category'] == 'Jewelry']
len(jewelry)

250

In [53]:
office = [product for product in data if product['category'] == 'Office Products']
len(office)

297

In [70]:
food = [product for product in data if product['category'] == 'Grocery And Gourmet Food']
len(food)

81

In [74]:
home_garden = [product for product in data if product['category'] == 'Home And Garden']

In [36]:
food[0]

{'id': 274490,
 'cluster_id': 141824.0,
 'category': 'Grocery And Gourmet Food',
 'input_title': 'Nutiva Organic Hemp Protein Hi-Fiber -- 3 lbs',
 'input_description': "Per Serving 12 g Fiber 11 g Protein Nutiva's delicious Organic Hemp Protein Powder is: Extremely nutritious. One serving provides 11 grams raw organic protein and a whopping 14 grams of fiber (54% of the RDI). Hemp contains all 8 essential amino acids with the bonus of good-for-you essential fatty acids. Certified Organic. No GMOs, pesticides, or solvents! Ever wonder why soy burgers and soymilk are organic yet soy protein isn't? Most non-organic soy protein is processed with hexane, a petroleum solvent similar to gasoline. Planet-friendly. Hemp (a non-drug crop) is the earth's premier renewable resource and grows easily without pesticides.",
 'target_scores': {'Brand': {'Nutiva': {'pid': [0, 1], 'score': 1}},
  'Product Type': {'Hemp Protein': {'pid': [0], 'score': 1},
   'Hemp Protein Powder': {'pid': [1], 'score': 1}

In [62]:
key_set = {}

attributes_list = []
for i in jewelry:
   attributes_list += list(i['target_scores'].keys())

attributes = set(attributes_list)
attributes

{'Brand',
 'Gender',
 'Metal Type',
 'Model Number',
 'Product Type',
 'Stone Shape',
 'Stones Setting',
 'Stones Total Weight',
 'Stones Type'}

In [63]:
from pydantic import BaseModel
from typing import Optional

class ProductInfoGarden(BaseModel):
   Base: Optional[str] = Field(default=None, alias='Base')
   Capacity: Optional[str] = Field(default=None, alias='Capacity')
   Color: Optional[str] = Field(default=None, alias='Color')
   Cooling: Optional[str] = Field(default=None, alias='Cooling')
   Depth: Optional[str] = Field(default=None, alias='Depth')
   Gauge: Optional[str] = Field(default=None, alias='Gauge')
   Heat: Optional[str] = Field(default=None, alias='Heat')
   Height: Optional[str] = Field(default=None, alias='Height')
   Length: Optional[str] = Field(default=None, alias='Length')
   Manufacturer: Optional[str] = Field(default=None, alias='Manufacturer')
   Manufacturer_Stock_Number: Optional[str] = Field(default=None, alias='Manufacturer Stock Number')
   Material: Optional[str] = Field(default=None, alias='Material')
   Product_Type: Optional[str] = Field(default=None, alias='Product Type')
   Retail_UPC: Optional[str] = Field(default=None, alias='Retail UPC')
   Shape: Optional[str] = Field(default=None, alias='Shape')
   Shelves: Optional[str] = Field(default=None, alias='Shelves')
   Splash: Optional[str] = Field(default=None, alias='Splash')
   Stainless_Steel_Series: Optional[str] = Field(default=None, alias='Stainless Steel Series')
   Voltage: Optional[str] = Field(default=None, alias='Voltage')
   Width: Optional[str] = Field(default=None, alias='Width')


class ProductInfoFood(BaseModel):

   Brand: Optional[str] = Field(default=None, alias='Brand')
   Flavor: Optional[str] = Field(default=None, alias='Flavor')
   Manufacturer_Stock_Number: Optional[str] = Field(default=None, alias='Manufacturer Stock Number')
   Pack_Quantity: Optional[str] = Field(default=None, alias='Pack Quantity')
   Packing_Type: Optional[str] = Field(default=None, alias='Packing Type')
   Product_Type: Optional[str] = Field(default=None, alias='Product Type')
   Retail_UPC: Optional[str] = Field(default=None, alias='Retail UPC')
   Size_Weight: Optional[str] = Field(default=None, alias='Size/Weight')
   
   
class ProductInfoOffice(BaseModel):
   Brand: Optional[str] = Field(default=None, alias='Brand')
   Binding: Optional[str] = Field(default=None, alias='Binding')
   Capacity: Optional[str] = Field(default=None, alias='Capacity')
   Closure: Optional[str] = Field(default=None, alias='Closure')
   Colors: Optional[str] = Field(default=None, alias='Color(s)')
   Depth: Optional[str] = Field(default=None, alias='Depth')
   Height: Optional[str] = Field(default=None, alias='Height')
   Length: Optional[str] = Field(default=None, alias='Length')
   Width: Optional[str] = Field(default=None, alias='Width')
   Mounting: Optional[str] = Field(default=None, alias='Mounting')
   Manufacturer_Stock_Number: Optional[str] = Field(default=None, alias='Manufacturer Stock Number')
   Material: Optional[str] = Field(default=None, alias='Material')
   Product_Type: Optional[str] = Field(default=None, alias='Product Type')
   Retail_UPC: Optional[str] = Field(default=None, alias='Retail UPC')
   Page_Yield: Optional[str] = Field(default=None, alias='Page Yield')
   Paper_Weight: Optional[str] = Field(default=None, alias='Paper Weight')
   Sheet_Perforation: Optional[str] = Field(default=None, alias='Sheet Perforation')
   Pack_Quantity: Optional[str] = Field(default=None, alias='Pack_Quantity')
   
class ProductInfoJewelry(BaseModel):
   Brand: Optional[str] = Field(default=None, alias='Brand')
   Gender: Optional[str] = Field(default=None, alias='Gender')
   Metal_Type: Optional[str] = Field(default=None, alias='Metal Type')
   Model_Number: Optional[str] = Field(default=None, alias='Model Number')
   Product_Type: Optional[str] = Field(default=None, alias='Product Type')
   Stone_Shape: Optional[str] = Field(default=None, alias='Stone Shape')
   Stones_Setting: Optional[str] = Field(default=None, alias='Stones Setting')
   Stones_Total_Weight: Optional[str] = Field(default=None, alias='Stones Total Weight')
   Stones_Type: Optional[str] = Field(default=None, alias='Stones Type')


 


In [75]:
def initial_extraction(text: str, custom_extraction_prompt:str, model_name: str = 'gpt-4o') -> ProductInfoGarden:
    
    # Patch the OpenAI client with Instructor
    client = instructor.from_openai(OpenAI(api_key=os.getenv('OPENAI_KEY')))
    
    system_message = """
    You are an intelligent text extraction and conversion assistant. Your task is to extract structured information 
    from the given text and convert it into a structured format. 
    The output response should contain only the data extracted from the text, with no additional commentary, explanations, or extraneous information.
    If the required information could not be found from the given source, return nothing for that field. Do not hallucinate.
    """
    
    rule_prompt = """
                Here are the rules that you need to adhere:
                    ## Rules:
                    - The aim is to achieve simplicity and clarity in the extracted text.
                    - Make sure to answer in the structured format.
                    - If no information is provided for any of the fields, return nothing of that field.
                    - DO NOT HALLUCINATE.
                """
    
    extraction_prompt = f"""
    {system_message}
    {custom_extraction_prompt}
    """

    response = client.chat.completions.create(
        model=model_name, 
        response_model=ProductInfoGarden,
        messages=[
            {"role": "system", "content": extraction_prompt},
            {"role": "user", "content": f"Use the given text to extract information: {text}"},
            {"role": "user", "content": rule_prompt}
        ]
    )
    return response

def evaluate_extraction(my_output, expected_output):
    
    # Simplify the expected output by removing the details of pid and score and handling multiple valid options
    simplified_expected_output = {'target_scores': {}}

    for field, values in expected_output['target_scores'].items():
        if isinstance(values, dict):  # Check if the value is a dictionary (implying multiple values)
            # Extract keys and make them a list if there are multiple valid responses
            simplified_expected_output['target_scores'][field] = list(values.keys())
        else:
            # Directly assign if it is a single value
            simplified_expected_output['target_scores'][field] = values
    
    target_scores = simplified_expected_output['target_scores']
    total_fields = len(target_scores)  # Total fields to be extracted
    correct_matches = 0  # Count of correctly extracted fields
    incorrect_fields = []  # List to store names of fields that are incorrect

    for field, expected_values in target_scores.items():
        # Normalize the value from my_output
        field = field.replace(' ', '_')
        field = field.replace('/', '_')
        field = field.replace('(', '')
        field = field.replace(')', '')
        my_value = my_output.get(field, 'n/a') if my_output.get(field) is not None else 'n/a'
        my_value = my_value.lower()  # Convert to lower case for case insensitive comparison

        # Handle cases where multiple correct answers are expected
        if isinstance(expected_values, list):
            # Normalize all expected values for case-insensitive comparison
            expected_values_normalized = [value.lower() for value in expected_values]
            if my_value in expected_values_normalized:
                correct_matches += 1
            else:
                incorrect_fields.append(field)
        else:  # Single correct answer case
            expected_value_normalized = expected_values.lower()
            if my_value == expected_value_normalized:
                correct_matches += 1
            else:
                incorrect_fields.append(field)

    accuracy_percentage = (correct_matches / total_fields) * 100

    # Print fields that were incorrect
    if incorrect_fields:
        print("Incorrect fields:")
        for field in incorrect_fields:
            print(f"- {field}")

    return accuracy_percentage




In [78]:
result_list = []
accuracy_list = []
model_name = 'gpt-4o-mini' #'gpt-3.5-turbo'
dataset = 'garden'

for index in tqdm(range(0, 100)):
    result = {}
    data = home_garden[index]
    result['id'] = data['id']
    result['cluster_id'] = data['cluster_id']
    result['category'] = data['category']
    text_compile = f'title: {data["input_title"]}\ndescription: {data["input_description"]}'

    custom_extraction_prompt = """
        Extract information from the text extracted from a webpage of a product, including title and description:
        Output in a structured format.
        """
    response = initial_extraction(text_compile, custom_extraction_prompt, model_name)
    result['output'] = response.dict()
    accuracy = evaluate_extraction(response.dict(), data)
    result['accuracy'] = accuracy
    result_list.append(result)
    accuracy_list.append(accuracy)
print(f'Average accuracy: {sum(accuracy_list)/len(accuracy_list)}')

# Specify the file path for your JSON Lines file
file_path = f'{model_name.replace(".", "-")}-{dataset}-zero.jsonl'

write_jsonl_file(file_path, result_list)



  1%|          | 1/100 [00:02<04:28,  2.71s/it]

Incorrect fields:
- Width
- Length
- Depth
- Gauge
- Material
- Stainless_Steel_Series
- Height
- Base


  2%|▏         | 2/100 [00:05<04:29,  2.75s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Depth
- Gauge
- Material
- Stainless_Steel_Series
- Height
- Manufacturer_Stock_Number


  3%|▎         | 3/100 [00:07<04:13,  2.62s/it]

Incorrect fields:
- Manufacturer
- Length
- Material
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Base


  4%|▍         | 4/100 [00:11<04:47,  3.00s/it]

Incorrect fields:
- Color
- Height
- Width
- Length
- Material
- Shelves
- Manufacturer_Stock_Number


  5%|▌         | 5/100 [00:14<04:31,  2.86s/it]

Incorrect fields:
- Height
- Width
- Length
- Material
- Shelves
- Manufacturer_Stock_Number
- Base
- Capacity


  6%|▌         | 6/100 [00:15<03:54,  2.49s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Height
- Width
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Shelves
- Splash
- Shape
- Color
- Retail_UPC
- Manufacturer_Stock_Number
- Heat
- Gauge
- Base
- Voltage


  7%|▋         | 7/100 [00:18<03:48,  2.45s/it]

Incorrect fields:
- Product_Type
- Width
- Length
- Base
- Manufacturer_Stock_Number


  8%|▊         | 8/100 [00:20<03:36,  2.35s/it]

Incorrect fields:
- Manufacturer
- Height
- Length
- Depth
- Base
- Material
- Cooling
- Shape
- Color
- Retail_UPC
- Manufacturer_Stock_Number
- Width
- Heat
- Shelves
- Voltage
- Capacity


  9%|▉         | 9/100 [00:24<04:35,  3.03s/it]

Incorrect fields:
- Manufacturer
- Height
- Length
- Gauge
- Stainless_Steel_Series
- Base
- Material
- Splash
- Manufacturer_Stock_Number
- Width


 10%|█         | 10/100 [00:27<04:32,  3.03s/it]

Incorrect fields:
- Gauge
- Stainless_Steel_Series
- Base
- Material
- Shelves


 11%|█         | 11/100 [00:31<04:42,  3.17s/it]

Incorrect fields:
- Manufacturer
- Height
- Length
- Depth
- Stainless_Steel_Series
- Base
- Material
- Cooling
- Shape
- Color
- Retail_UPC
- Manufacturer_Stock_Number
- Width
- Heat
- Shelves
- Voltage
- Capacity


 12%|█▏        | 12/100 [00:34<04:22,  2.98s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Stainless_Steel_Series
- Base
- Splash
- Manufacturer_Stock_Number


 13%|█▎        | 13/100 [00:36<04:17,  2.96s/it]

Incorrect fields:
- Manufacturer
- Stainless_Steel_Series
- Base
- Material
- Manufacturer_Stock_Number
- Shelves


 14%|█▍        | 14/100 [00:40<04:20,  3.03s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Height
- Width
- Material
- Length
- Base


 15%|█▌        | 15/100 [00:42<04:03,  2.86s/it]

Incorrect fields:
- Product_Type
- Material
- Base


 16%|█▌        | 16/100 [00:45<04:05,  2.92s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Length
- Depth
- Stainless_Steel_Series
- Splash
- Color
- Manufacturer_Stock_Number
- Gauge
- Material
- Base


 17%|█▋        | 17/100 [00:47<03:30,  2.53s/it]

Incorrect fields:
- Height
- Length
- Base
- Stainless_Steel_Series
- Width
- Gauge


 18%|█▊        | 18/100 [00:49<03:20,  2.45s/it]

Incorrect fields:
- Width
- Length
- Height
- Base
- Capacity


 19%|█▉        | 19/100 [00:51<03:01,  2.24s/it]

Incorrect fields:
- Product_Type
- Material
- Manufacturer_Stock_Number
- Length
- Base
- Capacity


 20%|██        | 20/100 [00:54<03:24,  2.56s/it]

Incorrect fields:
- Manufacturer
- Depth
- Height
- Base
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Material


 21%|██        | 21/100 [00:57<03:24,  2.59s/it]

Incorrect fields:
- Manufacturer
- Width
- Depth
- Height
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Color
- Material
- Capacity


 22%|██▏       | 22/100 [00:58<02:47,  2.14s/it]

Incorrect fields:
- Product_Type
- Color
- Base


 23%|██▎       | 23/100 [01:00<02:37,  2.05s/it]

Incorrect fields:
- Color
- Shape
- Retail_UPC
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Shelves
- Splash
- Height
- Width
- Heat
- Gauge
- Base


 24%|██▍       | 24/100 [01:01<02:26,  1.93s/it]

Incorrect fields:
- Product_Type
- Material
- Base
- Capacity


 25%|██▌       | 25/100 [01:03<02:11,  1.75s/it]

Incorrect fields:
- Product_Type
- Manufacturer
- Base


 26%|██▌       | 26/100 [01:04<02:09,  1.75s/it]

Incorrect fields:
- Product_Type
- Manufacturer
- Manufacturer_Stock_Number
- Material
- Capacity


 27%|██▋       | 27/100 [01:08<02:50,  2.34s/it]

Incorrect fields:
- Product_Type
- Width
- Base
- Heat
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Shelves
- Splash
- Shape
- Height
- Color
- Retail_UPC
- Manufacturer_Stock_Number
- Gauge
- Voltage
- Capacity


 28%|██▊       | 28/100 [01:11<02:54,  2.42s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Material
- Heat
- Manufacturer_Stock_Number


 29%|██▉       | 29/100 [01:13<02:47,  2.36s/it]

Incorrect fields:
- Height
- Width
- Retail_UPC
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Shelves
- Splash
- Shape
- Heat
- Base
- Voltage


 30%|███       | 30/100 [01:15<02:33,  2.19s/it]

Incorrect fields:
- Product_Type
- Material
- Manufacturer
- Color
- Base


 31%|███       | 31/100 [01:17<02:22,  2.07s/it]

Incorrect fields:
- Splash
- Shape
- Height
- Color
- Heat
- Gauge
- Base
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Retail_UPC
- Manufacturer_Stock_Number
- Width
- Shelves
- Voltage
- Capacity


 32%|███▏      | 32/100 [01:19<02:19,  2.05s/it]

Incorrect fields:
- Splash
- Height
- Length
- Stainless_Steel_Series
- Retail_UPC
- Width
- Shape
- Color
- Heat
- Gauge
- Base
- Depth
- Cooling
- Shelves
- Material
- Voltage
- Capacity


 33%|███▎      | 33/100 [01:21<02:21,  2.12s/it]

Incorrect fields:
- Material
- Base
- Capacity


 34%|███▍      | 34/100 [01:23<02:11,  2.00s/it]

Incorrect fields:
- Product_Type
- Manufacturer_Stock_Number
- Color
- Base
- Material
- Capacity


 35%|███▌      | 35/100 [01:24<01:56,  1.80s/it]

Incorrect fields:
- Manufacturer
- Manufacturer_Stock_Number
- Base
- Material


 36%|███▌      | 36/100 [01:25<01:51,  1.75s/it]

Incorrect fields:
- Product_Type
- Color
- Base
- Material


 37%|███▋      | 37/100 [01:27<01:46,  1.70s/it]

Incorrect fields:
- Product_Type
- Color
- Base


 38%|███▊      | 38/100 [01:29<01:58,  1.91s/it]

Incorrect fields:
- Splash
- Shape
- Height
- Color
- Heat
- Gauge
- Base
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Width
- Shelves
- Material
- Voltage


 39%|███▉      | 39/100 [01:32<02:04,  2.05s/it]

Incorrect fields:
- Manufacturer
- Base


 40%|████      | 40/100 [01:34<02:14,  2.23s/it]

Incorrect fields:
- Manufacturer_Stock_Number
- Color
- Base
- Material


 41%|████      | 41/100 [01:38<02:38,  2.69s/it]

Incorrect fields:
- Product_Type
- Material
- Manufacturer
- Base


 42%|████▏     | 42/100 [01:40<02:21,  2.44s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Manufacturer_Stock_Number
- Capacity


 43%|████▎     | 43/100 [01:42<02:16,  2.40s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Depth
- Splash
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Material
- Capacity


 44%|████▍     | 44/100 [01:44<02:06,  2.26s/it]

Incorrect fields:
- Product_Type
- Manufacturer


 45%|████▌     | 45/100 [01:47<02:04,  2.27s/it]

Incorrect fields:
- Heat
- Splash
- Shape
- Height
- Color
- Gauge
- Base
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Retail_UPC
- Manufacturer_Stock_Number
- Width
- Shelves
- Material
- Voltage
- Capacity


 46%|████▌     | 46/100 [01:49<01:58,  2.19s/it]

Incorrect fields:
- Manufacturer
- Material
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Color


 47%|████▋     | 47/100 [01:50<01:39,  1.87s/it]

Incorrect fields:
- Product_Type
- Material
- Base


 48%|████▊     | 48/100 [01:52<01:47,  2.07s/it]

Incorrect fields:
- Product_Type
- Base
- Material
- Capacity


 49%|████▉     | 49/100 [01:55<01:47,  2.12s/it]

Incorrect fields:
- Product_Type
- Base
- Manufacturer_Stock_Number
- Capacity


 50%|█████     | 50/100 [01:57<01:49,  2.20s/it]

Incorrect fields:
- Height
- Color
- Heat
- Base
- Manufacturer_Stock_Number
- Capacity


 51%|█████     | 51/100 [01:59<01:48,  2.21s/it]

Incorrect fields:
- Product_Type
- Splash
- Shape
- Height
- Heat
- Gauge
- Base
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Retail_UPC
- Manufacturer_Stock_Number
- Width
- Shelves


 52%|█████▏    | 52/100 [02:01<01:36,  2.01s/it]

Incorrect fields:
- Manufacturer
- Height
- Base
- Manufacturer_Stock_Number


 53%|█████▎    | 53/100 [02:03<01:42,  2.18s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Manufacturer_Stock_Number


 54%|█████▍    | 54/100 [02:06<01:46,  2.32s/it]

Incorrect fields:
- Depth
- Material
- Shelves
- Stainless_Steel_Series
- Cooling
- Splash
- Shape
- Retail_UPC
- Manufacturer_Stock_Number
- Heat
- Gauge
- Base
- Voltage
- Capacity


 55%|█████▌    | 55/100 [02:08<01:36,  2.15s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Color
- Base
- Material


 56%|█████▌    | 56/100 [02:09<01:26,  1.96s/it]

Incorrect fields:
- Manufacturer
- Length
- Gauge
- Stainless_Steel_Series
- Base
- Material
- Color
- Manufacturer_Stock_Number
- Capacity


 57%|█████▋    | 57/100 [02:13<01:42,  2.39s/it]

Incorrect fields:
- Gauge
- Stainless_Steel_Series
- Base
- Manufacturer_Stock_Number


 58%|█████▊    | 58/100 [02:16<01:54,  2.73s/it]

Incorrect fields:
- Manufacturer
- Stainless_Steel_Series
- Base
- Splash
- Manufacturer_Stock_Number


 59%|█████▉    | 59/100 [02:19<01:56,  2.85s/it]

Incorrect fields:
- Depth
- Stainless_Steel_Series
- Base
- Material
- Splash
- Cooling
- Shape
- Manufacturer_Stock_Number
- Heat
- Shelves
- Voltage


 60%|██████    | 60/100 [02:21<01:42,  2.56s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Length
- Depth
- Gauge
- Stainless_Steel_Series
- Base
- Material
- Manufacturer_Stock_Number
- Width


 61%|██████    | 61/100 [02:23<01:28,  2.28s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Length
- Gauge
- Stainless_Steel_Series
- Base
- Material
- Manufacturer_Stock_Number
- Width


 62%|██████▏   | 62/100 [02:24<01:19,  2.09s/it]

Incorrect fields:
- Length
- Depth
- Gauge
- Stainless_Steel_Series
- Base
- Splash
- Cooling
- Shape
- Height
- Color
- Retail_UPC
- Manufacturer_Stock_Number
- Width
- Heat
- Shelves
- Voltage
- Capacity


 63%|██████▎   | 63/100 [02:27<01:26,  2.34s/it]

Incorrect fields:
- Length
- Stainless_Steel_Series
- Base
- Material
- Shelves


 64%|██████▍   | 64/100 [02:29<01:21,  2.27s/it]

Incorrect fields:
- Length
- Depth
- Base
- Material
- Manufacturer_Stock_Number
- Width


 65%|██████▌   | 65/100 [02:31<01:14,  2.14s/it]

Incorrect fields:
- Product_Type
- Height
- Width
- Depth
- Length
- Stainless_Steel_Series
- Shelves
- Splash
- Shape
- Color
- Retail_UPC
- Manufacturer_Stock_Number
- Heat
- Gauge
- Material
- Base
- Capacity


 66%|██████▌   | 66/100 [02:33<01:05,  1.91s/it]

Incorrect fields:
- Manufacturer_Stock_Number
- Base


 67%|██████▋   | 67/100 [02:35<01:07,  2.05s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Splash
- Height
- Color
- Heat
- Gauge
- Base
- Length
- Stainless_Steel_Series
- Cooling
- Retail_UPC
- Manufacturer_Stock_Number
- Width
- Shelves
- Voltage


 68%|██████▊   | 68/100 [02:37<01:08,  2.14s/it]

Incorrect fields:
- Heat
- Base


 69%|██████▉   | 69/100 [02:42<01:28,  2.85s/it]

Incorrect fields:
- Cooling
- Splash
- Shape
- Color
- Base
- Retail_UPC
- Width
- Shelves
- Material
- Capacity


 70%|███████   | 70/100 [02:44<01:19,  2.64s/it]

Incorrect fields:
- Product_Type
- Width
- Depth
- Material
- Stainless_Steel_Series
- Height
- Manufacturer_Stock_Number
- Gauge


 71%|███████   | 71/100 [02:52<02:00,  4.17s/it]

Incorrect fields:
- Product_Type
- Height
- Width
- Length
- Material
- Shelves
- Manufacturer_Stock_Number


 72%|███████▏  | 72/100 [02:54<01:42,  3.67s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Depth
- Manufacturer_Stock_Number
- Base
- Material


 73%|███████▎  | 73/100 [02:56<01:21,  3.03s/it]

Incorrect fields:
- Color
- Width
- Length
- Base
- Capacity


 74%|███████▍  | 74/100 [02:58<01:15,  2.89s/it]

Incorrect fields:
- Manufacturer_Stock_Number
- Base
- Material
- Capacity


 75%|███████▌  | 75/100 [03:00<01:04,  2.59s/it]

Incorrect fields:
- Product_Type
- Width
- Depth
- Base
- Splash
- Length
- Stainless_Steel_Series
- Cooling
- Shelves
- Shape
- Height
- Color
- Retail_UPC
- Manufacturer_Stock_Number
- Heat
- Gauge
- Voltage
- Capacity


 76%|███████▌  | 76/100 [03:02<00:58,  2.44s/it]

Incorrect fields:
- Base
- Shelves


 77%|███████▋  | 77/100 [03:04<00:53,  2.33s/it]

Incorrect fields:
- Width
- Length
- Base
- Capacity


 78%|███████▊  | 78/100 [03:06<00:49,  2.24s/it]

Incorrect fields:
- Product_Type
- Width
- Base
- Heat
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Shelves
- Splash
- Shape
- Height
- Color
- Retail_UPC
- Gauge
- Voltage
- Capacity


 79%|███████▉  | 79/100 [03:08<00:43,  2.08s/it]

Incorrect fields:
- Product_Type
- Width
- Base
- Heat
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Shelves
- Splash
- Shape
- Height
- Color
- Retail_UPC
- Gauge
- Voltage
- Capacity


 80%|████████  | 80/100 [03:09<00:35,  1.77s/it]

Incorrect fields:
- Color
- Product_Type
- Manufacturer_Stock_Number
- Base
- Material


 81%|████████  | 81/100 [03:12<00:39,  2.08s/it]

Incorrect fields:
- Heat
- Splash
- Color
- Base
- Stainless_Steel_Series
- Manufacturer_Stock_Number


 82%|████████▏ | 82/100 [03:14<00:37,  2.08s/it]

Incorrect fields:
- Product_Type
- Width
- Heat
- Splash
- Shape
- Height
- Color
- Gauge
- Base
- Length
- Stainless_Steel_Series
- Depth
- Cooling
- Retail_UPC
- Shelves
- Material
- Voltage
- Capacity


 83%|████████▎ | 83/100 [03:16<00:35,  2.09s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Depth
- Splash
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Capacity


 84%|████████▍ | 84/100 [03:18<00:33,  2.07s/it]

Incorrect fields:
- Manufacturer
- Color
- Product_Type
- Material
- Manufacturer_Stock_Number
- Base


 85%|████████▌ | 85/100 [03:21<00:33,  2.22s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Manufacturer_Stock_Number


 86%|████████▌ | 86/100 [03:23<00:30,  2.20s/it]

Incorrect fields:
- Material
- Splash
- Base
- Length
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Width


 87%|████████▋ | 87/100 [03:25<00:28,  2.21s/it]

Incorrect fields:
- Manufacturer
- Manufacturer_Stock_Number


 88%|████████▊ | 88/100 [03:28<00:27,  2.32s/it]

Incorrect fields:
- Manufacturer_Stock_Number


 89%|████████▉ | 89/100 [03:30<00:24,  2.27s/it]

Incorrect fields:
- Stainless_Steel_Series
- Manufacturer_Stock_Number
- Base
- Capacity


 90%|█████████ | 90/100 [03:33<00:25,  2.56s/it]

Incorrect fields:
- Manufacturer
- Stainless_Steel_Series
- Color
- Manufacturer_Stock_Number
- Base
- Capacity


 91%|█████████ | 91/100 [03:37<00:26,  2.91s/it]

Incorrect fields:
- Manufacturer
- Product_Type
- Width
- Depth
- Splash
- Stainless_Steel_Series
- Manufacturer_Stock_Number


 92%|█████████▏| 92/100 [03:40<00:23,  2.93s/it]

Incorrect fields:
- Manufacturer
- Color
- Product_Type
- Height
- Width
- Length
- Material
- Shelves
- Manufacturer_Stock_Number


 93%|█████████▎| 93/100 [03:42<00:18,  2.60s/it]

Incorrect fields:
- Heat
- Gauge
- Base
- Cooling
- Manufacturer_Stock_Number
- Shelves
- Voltage
- Capacity


 94%|█████████▍| 94/100 [03:45<00:16,  2.78s/it]

Incorrect fields:
- Manufacturer
- Base
- Shelves
- Manufacturer_Stock_Number


 95%|█████████▌| 95/100 [03:47<00:12,  2.48s/it]

Incorrect fields:
- Material
- Manufacturer_Stock_Number
- Base


 96%|█████████▌| 96/100 [03:49<00:10,  2.55s/it]

Incorrect fields:
- Manufacturer
- Width
- Length
- Material
- Base


 97%|█████████▋| 97/100 [03:52<00:08,  2.67s/it]

Incorrect fields:
- Width
- Cooling
- Length
- Stainless_Steel_Series
- Depth
- Shelves
- Splash
- Shape
- Height
- Color
- Retail_UPC
- Heat
- Gauge
- Material
- Base
- Voltage


 98%|█████████▊| 98/100 [03:54<00:04,  2.37s/it]

Incorrect fields:
- Manufacturer
- Base


 99%|█████████▉| 99/100 [03:56<00:02,  2.40s/it]

Incorrect fields:
- Manufacturer_Stock_Number
- Base


100%|██████████| 100/100 [03:59<00:00,  2.39s/it]

Incorrect fields:
- Height
- Width
- Length
- Depth
- Stainless_Steel_Series
- Cooling
- Splash
- Shape
- Retail_UPC
- Manufacturer_Stock_Number
- Heat
- Gauge
- Base
- Voltage
- Capacity
Average accuracy: 61.4


In [83]:
len(home_garden)

356

- garden - gpt3.5 - zero - 75.65
- garden - gpt4o - zero - 81.2
- garden - gpt4o-mini - zero - 62.3, 61.4

- food - gpt3.5 - zero - 70.21604938271605
- food - gpt4o - zero - 77.62345679012346, 75.92592592592592
- food - gpt4o-mini - zero - 74.07407407407408

- office - gpt3.5 - zero - 76.9444444444444
- office - gpt4o - zero - 78.11111111111109
- office - gpt4o-mini - zero - 76.05555555555554

- jewelry - gpt3.5 - zero - 76.77777777777769
- jewelry - gpt4o - zero - 84.55555555555544
- jewelry - gpt4o-mini - zero - 82.3333333333332





In [ ]:
Example = """
        Here is an example of the information extraction tasks
        
        ## Input Text
        title: John Boos EBMS6-2024 24" Solid Wall Mounted Shelving w/ Raised Ledges Ledges
        description: Microwave Shelf, wall mount, 24"W x 20"D x 15"H, 16 gauge 300 series stainless steel, Stallion safety edge front with raised lip, NSF
        
        ## Expected Output
        'Manufacturer': 'John Boos',
        'Product Type': 'Wall Mounted Shelving'
        'Width': '24"'
        'Length': None
        'Depth': '20"'
        'Gauge': '16'
        'Material': 'stainless steel'
        'Stainless Steel Series': '300'
        'Cooling': None 
        'Splash': None
        'Shape': None
        'Height': '15"'
        'Color': None
        'Retail UPC': None
        'Manufacturer Stock Number': 'EBMS6-2024'
        'Heat': None
        'Shelves': None
        'Base': None
        'Voltage': None
        'Capacity': None
    """

In [18]:
a = [1,2,3]


2.0